decal system & crayons

PaulRitter opened the pull request:
now also contains a rewrite of crayons to this system!
for some reason the cleaning-part does not work, hmu if you know why

does not yet have
- any mapping placement ui whatsoever
- rotation
- map-relative decals

SweptWasTaken commented:
Have you done any performance testing with the decal system to see if it would support the large amounts of decals on SS13 maps?

Do decals have a toggle to be examinable? It looks like you removed the examine text from the original crayon stuff. 

Also should signs be converted to the decal system since they are on SS13 codebase? If you can't give them a name and description yet then they shouldn't be.

mirrorcult commented:
signs def should just be entities
making them examinable is probably harder than it seems since tiles aren't examinable either

PaulRitter commented:
> 
> 
> Have you done any performance testing with the decal system to see if it would support the large amounts of decals on SS13 maps?
> 
> Do decals have a toggle to be examinable? It looks like you removed the examine text from the original crayon stuff.
> 
> Also should signs be converted to the decal system since they are on SS13 codebase? If you can't give them a name and description yet then they shouldn't be.

Perf will improve since you dont have any overhead by them being entities. I also implemented chunking & delta updates to further improve perf. If you want to run a benchmark, be my guest.

No, decals dont enable to be examinable since they arent entities.

Signs should be entities.

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,227 @@
+            if (_removedDecals.Count > 0)
+                RaiseNetworkEvent(new DecalRemovalUpdateEvent{RemovedDecals = _removedDecals});
+
+            foreach (var playerSession in _playerManager.GetAllPlayers())
+            {
+                if (!_previousSentChunks.TryGetValue(playerSession, out var previous))
+                    previous = new();
```
Just listen to the PlayerStatusChange event and then you can also clean these up as people disconnect.

metalgearsloth commented on Content.Shared/Decals/Decal.cs:
```diff
@@ -0,0 +1,14 @@
+namespace Content.Shared.Decals
+{
+    [Serializable, NetSerializable]
+    public record Decal(Vector2 Coordinates, string Id, Color? Color, Angle Angle, int ZIndex)
```
based

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,227 @@
+        {
+            base.Update(frameTime);
+
+            if (_removedDecals.Count > 0)
+                RaiseNetworkEvent(new DecalRemovalUpdateEvent{RemovedDecals = _removedDecals});
```
Probably a bad idea for bandwidth reasons?

PaulRitter replied:
how come? i guess i could tie it into the other updatemessage

metalgearsloth replied:
Say decals get changed on salvage; someone who's on the station doesn't probably care about whether they're removed or not as they've never seen it. Though I guess the trade off is that you need to track what chunks each player has seen.

PaulRitter replied:
> Though I guess the trade off is that you need to track what chunks each player has seen.

not really. if i were to send this in the updatemessage, we would update all seen chunks by the player anyways. we dont really care if the data outside the viewport is incorrect

metalgearsloth commented on Content.Shared/Decals/SharedDecalSystem.cs:
```diff
@@ -0,0 +1,207 @@
+            SubscribeLocalEvent<GridInitializeEvent>(OnGridInitialize);
+            SubscribeLocalEvent<GridRemovalEvent>(OnGridRemoval);
+            _viewSize = _configurationManager.GetCVar(CVars.NetMaxUpdateRange)*2f;
+            _configurationManager.OnValueChanged(CVars.NetMaxUpdateRange, f => _viewSize = f*2f);
```
Need to unsub.

PaulRitter commented on Content.Shared/Decals/SharedDecalSystem.cs:
```diff
@@ -0,0 +1,207 @@
+
+        private float _viewSize;
+
+        public override void Initialize()
+        {
+            base.Initialize();
+
+            SubscribeLocalEvent<GridInitializeEvent>(OnGridInitialize);
+            SubscribeLocalEvent<GridRemovalEvent>(OnGridRemoval);
+            _viewSize = _configurationManager.GetCVar(CVars.NetMaxUpdateRange)*2f;
+            _configurationManager.OnValueChanged(CVars.NetMaxUpdateRange, f => _viewSize = f*2f);
+        }
```
i was planning on subscribing to tiles being removed anyways, removing any decals on them. would this also fix it?

metalgearsloth replied:
Yeah it should

metalgearsloth commented on Content.Shared/Decals/SharedDecalSystem.cs:
```diff
@@ -0,0 +1,153 @@
+            base.Initialize();
+
+            SubscribeLocalEvent<GridInitializeEvent>(OnGridInitialize);
+            _viewSize = _configurationManager.GetCVar(CVars.NetMaxUpdateRange)*2f;
+            _configurationManager.OnValueChanged(CVars.NetMaxUpdateRange, OnPvsRangeChanged);
```
you can just pass in true as the last arg and it gets invoked immediately.

metalgearsloth commented on Content.Shared/Decals/SharedDecalSystem.cs:
```diff
@@ -0,0 +1,153 @@
+            foreach (var viewerUid in viewers)
+            {
+                var (bounds, mapId) = CalcViewBounds(viewerUid);
+                foreach (var grid in MapManager.FindGridsIntersecting(mapId, bounds, approx: true))
+                {
+                    if(!chunks.ContainsKey(grid.Index))
+                        chunks[grid.Index] = new();
+                    var enumerator = new ChunkIndicesEnumerator(grid.InvWorldMatrix.TransformBox(bounds), ChunkSize);
+                    while (enumerator.MoveNext(out var indices))
+                    {
+                        chunks[grid.Index].Add(indices.Value);
+                    }
+                }
+            }
```
There's an enumerator for FindGridsIntersecting if you want to get fancy

PaulRitter replied:
what are the upsides of using an enumerator like that?

metalgearsloth replied:
Allocations

metalgearsloth commented on Content.Client/Decals/DecalOverlay.cs:
```diff
@@ -0,0 +1,60 @@
+                return spriteSpecifier;
+            }
+
+            foreach (var (gridId, zIndexDictionary) in _system.DecalRenderIndex)
+            {
+                var grid = _mapManager.GetGrid(gridId);
```
When it's a comp you can just get OwnerUid + the transform directly.

metalgearsloth commented on Content.Client/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,102 @@
+        {
+            base.Initialize();
+
+            _overlay = new DecalOverlay(this, MapManager, PrototypeManager);
+            _overlayManager.AddOverlay(_overlay);
```
Future concern but would be nice to have a console command to turn off decal rendering (for mapping purposes or otherwise).

metalgearsloth commented on Content.Server/Crayon/CrayonComponent.cs:
```diff
@@ -57,10 +60,10 @@ protected override void Initialize()
             Charges = Capacity;
 
             // Get the first one from the catalog and set it as default
-            var decals = _prototypeManager.EnumeratePrototypes<CrayonDecalPrototype>().FirstOrDefault();
-            if (decals != null)
+            var decals = _prototypeManager.EnumeratePrototypes<DecalPrototype>().Where(x => x.Tags.Contains("crayon")).ToArray();
+            if (decals.Length != 0)
             {
-                SelectedState = decals.Decals.First();
+                SelectedState = decals[0].ID;
```
Dear lord I hope someone ECS this soon.

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,277 @@
+        {
+            switch (e.NewStatus)
+            {
+                case SessionStatus.Connected:
+                    _previousSentChunks[e.Session] = new();
+                    break;
```
PVS uses InGame though I'm not sure what the difference is in terms of networking.

I'd probably look at using Connected / InGame as the same one and having default as remove.

PaulRitter replied:
i should use ingame. its when the player actually exists the lobby and plays the game.

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,277 @@
+        private HashSet<EntityUid> GetSessionViewers(IPlayerSession session)
+        {
+            var viewers = new HashSet<EntityUid>();
+            if (session.Status != SessionStatus.InGame || session.AttachedEntityUid is null)
+                return viewers;
```
Elsewhere you check for Connected but this checks for InGame.

PaulRitter replied:
ye as i said above, only need to send it to them when they are actually ingame, so i should also only create the index for players that are ingame

PaulRitter commented on Content.Shared/Decals/Decal.cs:
```diff
@@ -0,0 +1,31 @@
+{
+    [Serializable, NetSerializable]
+    [DataDefinition]
+    public readonly struct Decal
```
suggestions what i could change about it?

PaulRitter commented on Content.Server/Chemistry/TileReactions/CleanTileReaction.cs:
```diff
@@ -32,6 +37,14 @@ FixedPoint2 ITileReaction.TileReact(TileRef tile, ReagentPrototype reagent, Fixe
+            var decalSystem = EntitySystem.Get<DecalSystem>();
+            var prototypeManager = IoCManager.Resolve<IPrototypeManager>();
+            foreach (var uid in decalSystem.GetDecalsOnTile(tile.GridIndex, tile.GridIndices,
+                x => prototypeManager.TryIndex<DecalPrototype>(x.Id, out var decal) && decal.Tags.Contains("crayon")))
```
did that initially, dunno why i reverted that. will fix

PaulRitter replied:
ah, i remembered. it doesnt really make sense to specify these things on the prototype itself, so i was planning on adding a tags variable (or a cleanable flag) to the actual decal object

PaulRitter commented on Content.Client/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,87 @@
+        [Dependency] private readonly IOverlayManager _overlayManager = default!;
+
+        private DecalOverlay _overlay = default!;
+        public Dictionary<GridId, SortedDictionary<int, SortedDictionary<uint, Decal>>> DecalRenderIndex = new();
```
this is already a component on the serverside. these indices serve purely for rendering and get updated/populated on receiving messages from the server

PaulRitter commented:
i addressed the reviews, however:
- i still need to test editing
- in my short test ingame, rendering didnt work anymore after my fixes. i'll have to look into that
- still need to test overlay toggling

metalgearsloth commented on Content.Client/Decals/DecalOverlay.cs:
```diff
@@ -0,0 +1,60 @@
+                    foreach (var (_, decal) in decals)
+                    {
+                        var spriteSpecifier = GetSpriteSpecifier(decal.Id);
+                        handle.DrawTexture(spriteSpecifier.Frame0(), decal.Coordinates, decal.Color);
```
Update to new method.

metalgearsloth commented on Content.Client/Decals/DecalOverlay.cs:
```diff
@@ -0,0 +1,60 @@
+                return spriteSpecifier;
+            }
+
+            foreach (var (gridId, zIndexDictionary) in _system.DecalRenderIndex)
+            {
+                var grid = _mapManager.GetGrid(gridId);
+                handle.SetTransform(grid.WorldMatrix);
```
Get grident and then transform (someday mapgrid will wholly be a component TM).

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,323 @@
+            DirtyChunk(e.NewTile.GridIndex, indices);
+        }
+
+        private void OnPlayerStatusChanged(object? sender, SessionStatusEventArgs e)
+        {
+            switch (e.NewStatus)
+            {
+                case SessionStatus.InGame:
+                    _previousSentChunks[e.Session] = new();
+                    break;
+                case SessionStatus.Disconnected:
+                    _previousSentChunks.Remove(e.Session);
+                    break;
+            }
+        }
```
Maybe make the default to remove instead?

PaulRitter replied:
probably a good idea, yeah

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,323 @@
+            if (MapManager.GetGrid(gridId).GetTileRef(coordinates).IsSpace())
+                return false;
+
+            rotation ??= Angle.Zero;
+            var decal = new Decal(coordinates.Position, id, color, rotation.Value, zIndex, cleanable);
+            var chunkCollection = DecalGridChunkCollection(gridId);
+            uid = chunkCollection.NextUid++;
+            var chunkIndices = GetChunkIndices(decal.Coordinates);
+            if(!chunkCollection.ChunkCollection.ContainsKey(chunkIndices))
+                chunkCollection.ChunkCollection[chunkIndices] = new();
+            chunkCollection.ChunkCollection[chunkIndices][uid.Value] = decal;
+            ChunkIndex[gridId][uid.Value] = chunkIndices;
+            DirtyChunk(gridId, chunkIndices);
+            return true;
```
holy wall of text man

PaulRitter replied:
we do a little chunkchunkchunkchunkuiddecal

metalgearsloth commented on Content.Server/Decals/DecalSystem.cs:
```diff
@@ -0,0 +1,323 @@
+        {
+            base.Update(frameTime);
+
+            foreach (var playerSession in _playerManager.GetAllPlayers())
+            {
+                var chunks = GetChunksForSession(playerSession);
```
Do the warning.
Also if you check for InGame here it avoids needing to allocate updatedChunks before realising they're not ingame.

PaulRitter replied:
dang, totally missed that. what do you mean by "do the warning"?

metalgearsloth commented on Content.Shared/Decals/SharedDecalSystem.cs:
```diff
@@ -0,0 +1,153 @@
+
+        private void OnGridInitialize(GridInitializeEvent msg)
+        {
+            var comp = EntityManager.EnsureComponent<DecalGridComponent>(MapManager.GetGrid(msg.GridId).GridEntityId);
+            ChunkIndex[msg.GridId] = new();
```
The mix of dictionary and component activates my almonds.

PaulRitter replied:
sir its called an index. that being said this one doesnt get rebuilt after loading so i should probably do that
